lightning/src/ln/chan_utils.rs

```diff
@@ -690,25 +764,317 @@ impl HolderCommitmentTransaction {
+	}
+
+	/// Build the Bitcoin transaction.
+	///
+	/// The second return value may be used to debug the script that were built, since
+	/// only the hash is present in the actual transaction.
+	pub fn build<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Result<(BuiltCommitmentTransaction, Vec<Script>), ()> {
+		let (obscured_commitment_transaction_number, txins) = Self::internal_build_inputs(self.commitment_number, channel_parameters);
+
+		let mut txouts = self.build_outputs(channel_parameters, secp_ctx)?;
+
+		let mut outputs = Vec::with_capacity(txouts.len());
+		let mut scripts = Vec::with_capacity(txouts.len());
+		for (txout, script) in txouts.drain(..) {
+			outputs.push(txout);
+			scripts.push(script);
+		}
+
+		let transaction = Self::make_transaction(obscured_commitment_transaction_number, txins, outputs);
+		let txid = transaction.txid();
+		let built_transaction = BuiltCommitmentTransaction {
+			transaction,
+			txid
+		};
+		Ok((built_transaction, scripts))
+	}
+
+	fn make_transaction(obscured_commitment_transaction_number: u64, txins: Vec<TxIn>, outputs: Vec<TxOut>) -> Transaction {
+		Transaction {
+			version: 2,
+			lock_time: ((0x20 as u32) << 8 * 3) | ((obscured_commitment_transaction_number & 0xffffffu64) as u32),
+			input: txins,
+			output: outputs,
+		}
+	}
+
+	fn build_outputs<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Result<Vec<(TxOut, Script)>, ()> {
+		let mut aux = Vec::with_capacity(self.htlcs.len());
+		aux.resize(self.htlcs.len(), ());
+		let mut txouts = Self::internal_build_outputs(&self.keys, self.to_broadcaster_value_sat, self.to_countersignatory_value_sat, &self.htlcs, aux, channel_parameters, secp_ctx)?;
+		let outs = txouts.drain(..).map(|(out, (s, _))| (out, s)).collect();
+		Ok(outs)
+	}
+
+	// This is used in two cases:
+	// - initial sorting of outputs / HTLCs in the constructor, in which case T is auxiliary data the
+	//   caller needs to have sorted together with the HTLCs so it can keep track of the output index
+	// - building of a bitcoin transaction (build -> build_outputs), in which case T is just ()
+	fn internal_build_outputs<T, S: secp256k1::Signing + secp256k1::Verification>(keys: &TxCreationKeys, to_broadcaster_value_sat: u64, to_countersignatory_value_sat: u64, htlcs: &Vec<HTLCOutputInCommitment>, aux: Vec<T>, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<S>) -> Result<Vec<(TxOut, (Script, Option<(HTLCOutputInCommitment, T)>))>, ()> {
+		assert_eq!(htlcs.len(), aux.len(), "htlcs vs aux length mismatch");
+		let (broadcaster_pubkeys, countersignatory_pubkeys) = channel_parameters.pubkeys();
+		let contest_delay = channel_parameters.contest_delay();
+
+		let per_commitment_point = &keys.per_commitment_point;
+		let to_broadcaster_delayed_pubkey = derive_public_key(
```

Comment:
Hmm, can we avoid deriving all these keys twice (once when we create the TxCreationKeys object and once when we get here, eg from `Channel::build_commitment_transaction`)? Obviously would be nice if we could do it only here and not in `Channel` at all, but if not maybe we should have a version here that implies trusting `self.keys` and one that verifies `self.keys` is correct first? Same goes for revokeable_redeemscript below.

Reply:
I'm now passing in the keys in `build()` and also added docs there to clarify that trusted keys must be passed in.

I'm not sure what you mean about `revokeable_redeemscript` since it's just putting some bytes in a buffer so should be cheap.